### PR TITLE
feat: clean up and simplify logging functions and add support for printing classes

### PR DESCRIPTION
**Pull request by Enduriel**

### CODE REVIEW

File: scripts/config/msu/log.nut [hunk 1]
```diff
@@ -25,7 +25,7 @@
 					fixCounter++;
 					continue;
 				}
-				string += this.getLocalString(key, value, _maxLen, _maxDepth, _advanced);
+				string += key + " = " + this.getLocalString(value, _maxLen, _maxDepth, _advanced, false) + ", ";
```
Comment (TaroEld) [inline scripts/config/msu/log.nut:28]: Use format()

File: scripts/config/msu/log.nut [hunk 2]
```diff
@@ -36,60 +36,78 @@
...
-	function formatData( _data, _maxDepth = 1, _advanced = false, _maxLenMin = 1)
+	function printClass( _data, _maxLen = 128, _maxDepth = 1, _advanced = false )
 	{
-		if ((typeof _data == "array" || typeof _data == "table") && _data.len() > _maxLenMin)
+		this.printData(_data, _maxDepth, _advanced, _maxLen, true);
+	}
+
+	function formatData( _data, _maxDepth = 1, _advanced = false, _maxLenMin = 1, _printClasses = false)
+	{
+		if (["array", "table"].find(typeof _data) != null  && _data.len() > _maxLenMin)
```
Comment (TaroEld) [inline scripts/config/msu/log.nut:51]: I prefer `if (typeof _data == ("array" || "table"))`

File: scripts/config/msu/log.nut [hunk 3]
```diff
@@ -36,60 +36,78 @@
...
-				{
-					string += ::MSU.String.capitalizeFirst(typeof _value) + ", ";
-					break;
-				}
-			default:
-				string += _value + ", ";
+			ret += "{";
+			foreach (key, value in _value)
+			{
+				ret += key + " = " + this.getLocalString(value, _maxLen, _depth - 1, _advanced, _printClasses) + ", ";
```
Comment (TaroEld) [inline scripts/config/msu/log.nut:77]: Use format()

File: scripts/config/msu/log.nut [hunk 4]
```diff
@@ -36,60 +36,78 @@
...
+			ret += "}";
+		}
+		else if (["instance", "class"].find(typeof _value) != null && _depth > 0 && _printClasses) // full instance or class
+		{
+			ret += "&lt;"; // < in log
+			local valueClass = typeof _value == "instance" ? _value.getclass() : _value;
+			local len = 0;
+			foreach (key, value in valueClass)
+			{
+				ret += key + " = " + this.getLocalString(_value[key], _maxLen, _depth - 1, _advanced, _printClasses) + ", ";
```
Comment (TaroEld) [inline scripts/config/msu/log.nut:89]: Use format()

File: scripts/config/msu/log.nut [hunk 5]
```diff
@@ -36,60 +36,78 @@
...
+			if (len != 0) ret = ret.slice(0, -2);
+			ret += "&gt;"; // > in log
+		}
+		else if (!_advanced && ["function", "instance", "table", "array", "null", "class"].find(typeof _value) != null)
+		{
+			ret += ::MSU.String.capitalizeFirst(typeof _value);
+		}
+		else if (typeof _value == "string")
+		{
+			ret += "\"" + _value + "\"";
```
Comment (TaroEld) [inline scripts/config/msu/log.nut:101]: Use format()

**Review by TaroEld — COMMENTED** [review]
Please use format when concatenating more than 2 strings

**Review by LordMidas — CHANGES_REQUESTED** [review]

File: scripts/config/msu/log.nut [hunk 6]
```diff
@@ -36,60 +36,78 @@
...
+			foreach (idx, value in _value)
+			{
+				ret += this.getLocalString(value, _maxLen, _depth - 1, _advanced, _printClasses) + ", ";
+			}
+			if (_value.len() != 0) ret = ret.slice(0, -2);
+			ret += "]";
 		}
-		local arrayVsTable = ["{", false, "}"];
-		switch (typeof _value)
+		else if (typeof _value == "table" && _value.len() <= _maxLen && _depth > 0) // full table
```
[the remark this answers is not in the document]
Comment (Enduriel) [reply scripts/config/msu/log.nut:72]: This was originally a switch as you can see, it's not actually cleaner and makes it a lot more difficult to parse, a chain of else ifs makes it a little annoying, but is MUCH better than the alternative.

File: scripts/config/msu/log.nut [hunk 7]
```diff
@@ -36,60 +36,78 @@
...
-				string += _value + ", ";
+			ret += "{";
+			foreach (key, value in _value)
+			{
+				ret += format("%s = %s, ", key, this.getLocalString(value, _maxLen, _depth - 1, _advanced, _printClasses));
+			}
+			if (_value.len() != 0) ret = ret.slice(0, -2);
+			ret += "}";
+		}
+		else if (["instance", "class"].find(typeof _value) != null && _depth > 0 && _printClasses) // full instance or class
```
[the remark this answers is not in the document]
Comment (Enduriel) [reply scripts/config/msu/log.nut:82]: Refer to the other discussion

**Comment by Enduriel** [comment]
after discussion midas was okay with it as is

**Pull request merged**